### ISSUE TITLE
Scheduling timing issue

### ISSUE THREAD

**Issue by aralex1** [zephyrproject-rtos/zephyr#46656]
Hi everyone, my teammates and I are having a scheduling issue with our implementation. We have 3 K_ESSENTIAL threads with different priorities.
The highest priority thread (Thigh) has the numerical priority 2. 
The middle priority thread (Tmid) has the numerical priority 7. 
The lowest priority thread (Tlow) has the numerical priority 12.

The only thread that is allowed to do blocking calls is Tlow since the other ones can preempt it. The threads use QP with individual queues for messages. When there are no messages, the threads wait/sleep using a semaphore. 

We have a function on Tlow that wakes up Tmid by giving the semaphore and then it spins on a while loop while waiting for Tmid to finish its execution and set the flag that Tlow is waiting on. Once Tmid finishes, it goes back to wait/sleep by taking the semaphore. Thus, this should preempt Tlow once again since it is spinning. 

Tlow
   - wakes Tmed
   - spins waiting for the flag from Tmed

Tmed
   - does work
   - sets the flag when done
   - waits on semaphore

This process should take approx 300 microseconds. When we call this function 1 time, it works perfectly and takes the mentioned time. However, if we loop 100 times this goes from an expected ~0.05 seconds execution to a ~3 seconds execution. After an exhausting investigation, we figure out that using the k_yield() before spinning on Tlow gives us the expected result. As per the scheduling documentation ([Link](https://docs.zephyrproject.org/3.0.0/reference/kernel/scheduling/index.html#:~:text=transition%20of%20a,thread%20invokes%20k_yield())), we should obtain the same functionality by using a semaphore. Or, is expected to use the k_yield() in this type of scenario?

**Comment by DariusBabrauskas** [comment]
Hi. 
What you mean "spins waiting for the flag from Tmed"? 
Is this flag is kernel object or global variable?
What is QP?  It's kernel object?
Better part of code.

**Comment by andyross** [comment]
Sorry, just now noticing this had been submitted here.  What you describe should definitely work, can you trim it down and submit code that exercises it?  If I had to guess, what's happening is that there is confusion about what the thread priorities are (e.g. the low priority thread might be cooperative, or as @peter-mitsis suggested in discord might have had its priority boosted by the use of k_mutex).

**Comment by aralex1** [comment]
Hi, 
@DariusBabrauskas, It is a global flag that is only controlled/modified by Tmed. Tlow only reads it spinning on a while loop. 
QP is Quantum Leaps. I should have been more specific with this. The only reason why I mentioned this is because the system uses QP for its Active Objects and peripheral HSMs and the ideology of using messages to communicate between them. 

@andyross, I am sorry if I should have not created the ticket. I started the conversation on Slack and I thought that I needed to create a more official request over here. I am still getting used to the open-source community. So, I am not sure what is the expected procedure for discussing encountered problems. 
I will try to replicate the issue in a smaller scenario that I can upload here. If it also fails, I will share it over here; if not, I will close the ticket and take a closer look at our implementation. Thank you for your suggestion.

**Comment by andyross** [comment]
Creating bugs is fine!  But if you do that, you're sort of honor-bound to help us out with exerciser code, etc...  No reason not to leave this open for a bit, just keep us updated.

**Comment by mmahadevan108** [comment]
cc @peter-mitsis

**Comment by DariusBabrauskas** [comment]
Maybe need try to replace flag global variable with binary semaphore.

**Comment by aralex1** [comment]
@DariusBabrauskas thank you for the suggestion. We did try that approach and it worked by providing the expected timing result. However, this solution will be a bandage and it doesn't tell us if we are having a scheduling issue with our implementation and that's why we are still digging into it.

**Comment by peter-mitsis** [comment]
@aralex1 - Would you happen to have a version that replicates the problem that you can share? Even a stripped down version would be welcome. It would greatly help to narrow down the set of possible errors.

**Comment by aralex1** [comment]
@mmahadevan108, Apologize for the late response. 3 weeks ago, I was working on creating a smaller sample to try to recreate the issue, so I could share it here. However, I haven't been able to reproduce it, apart from our actual project. I was pulled 2 weeks ago to work on another project due to a higher priority and that's why I haven't had the chance to continue researching this issue.

**Comment by mmahadevan108** [comment]
@aralex1 any support you can give on this issue will be most helpful.

**Comment by aralex1** [comment]
@mmahadevan108 Like I mentioned, I haven't been able to go back to this research. But, here is what I tried to recreate the issue.
[TestCode.zip](https://github.com/zephyrproject-rtos/zephyr/files/9160038/TestCode.zip)

The idea is to have 2 threads running (Tight and Tlow). Thigh waits for Tlow on a semaphore. Once Tlow gives the semaphore to Thigh, Thigh does "some work" and then sleeps for 100ms. The sleeping is to allow Tlow to actually spin. This process repeats 100 times. 

If I remove the sleeping part, the code executes on 10 seconds. On the other hand, if I run it with sleep, it executes in 30 seconds. Since, I set CONFIG_SYS_CLOCK_TICKS_PER_SEC=10, K_MSEC(1) adds 100ms per turn. Thus, 10 seconds are added because of sleep, and switching between Tlow and Thigh is adding another 10 seconds. This tells me that the switching from spinning to Thigh takes an entire tick to actually happens. 

Note, I also have another test case in this sample code "TEST_SEM_FLAG" which changes the spinning for a semaphore. I didn't notice a difference in the timing by using the semaphore instead. This means that the semaphore also waits for the next tick to switch to Thigh. 

I used an nRF52840 DK board to execute this sample code.

**Comment by andyross** [comment]
I'm not able to reproduce, I don't think.  I built your sample code[1] with the only modification being to undefine TEST_SEM_FLAG, which I think is the failure case you're seeing.  And it's running in exactly 30s on all of the uniprocessor emulation environments I tried.

And some added printks shows that the two preemption points are working as expected:

1. When Tlow gives the semaphore, Thigh becomes runnable and immediately preempts Tlow as expected.
2. When Thigh becomes runnable again at the end of its sleep, the interrupt correctly returns into Thigh and not Tlow which was preempted.

Everything looks fine, basically.  Can you check again, and maybe share your .config file to see if there's anything else weird going on?

[1] Windows line endings and all.  Heh, which editor are you using that still does this?

**Comment by nashif** [comment]
does not seem to be high priority bug, can't be reproduced, not even by the original author.

**Comment by mmahadevan108** [comment]
@aralex1 Please reopen if you can reproduce the failure.
